keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,426 @@
+
+We will introduce two new metrics for this feature. Please refer to the "SLIs" section below for their specific meanings.
+If either the value of metric `pod_deletions_node_affinity_total` or `pod_deletions_node_affinity_duration_seconds` is larger than expected, users should check their system and may need to rollback.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Will be tested manually.
+
+Steps:
+- Create a node that has a label
+- Start a local Kubernetes cluster (feature gate defaulted disabled)
+- Create a Pod `test-pod` with `RequiredDuringSchedulingRequiredDuringExecution` node affinity matches the node
+- Check Pod's `requiredDuringSchedulingRequiredDuringExecution` node affinty gets dropped due to disabled feature gate
+- Delete the Pod
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Create `test-pod` again.
+- Verify that the pod is correctly scheduled to the node
+- Remove the node label
+- Verify that the pod is evicted
+- Add the node's label back
+- Create `test-pod` again
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the pod is evicted 
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Inspect the `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` configuration.
+- Check the corresponding metrics(details in SLIs part below).
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Events
+  - Event Reason: NodeAffinityEviction
+- [x] API .status
+  - Condition reason: DeletionByNodeAffinityEvictionController 
+- [x] Other (treat as last resort)
+  - Details: Pod with `RequiredDuringSchedulingRequiredDuringExecution` must meet affinity rule during both the scheduling phase and the execution phase.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name:
+    - pod_deletions_node_affinity_total(counts the number of Pods deleted because of nodeAffinity since this contoller start)
+    - pod_deletions_node_affinity_duration_seconds(latency, in seconds, between the time when a node's label has been changed and the pod been deleted via node-affinity-eviction Controller)
```

Comment:
Let's remove this metric. It might cost some engineering just to get a reasonable value, and it will not be a reliable number anyway.